egs/wsj/s5/steps/nnet3/cldnn/make_configs.py

```diff
@@ -0,0 +1,288 @@
+    
+    ## Work out cnn_array e.g. cnn_array = [ [8,8,1,1,256,1,3,1,1,3,1], [4,3,1,1,256,1,3,1,1,3,1 ] ]
+    cnn_array = []
+    split1 = args.cnn_indexes.split(" ");  # we already checked the string is nonempty.
+    if len(split1) < 1:
+      sys.exit("invalid --cnn-indexes argument, too short: "
+             + args.cnn_indexes)
+    try:
+      for string in split1:
+        split2 = string.split(",")
+        if len(split2) != 11:
+            sys.exit("invalid --cnn-indexes argument, must contain 11 positive integers: "
+                     + args.cnn_indexes)
+        int_list = []
+        for int_str in split2:
+            if int(int_str) < 1:
+                sys.exit("non-positive number in --cnn-indexes argument: {0}".format(int(int_str)))
+            int_list.append(int(int_str))
+        cnn_array.append(int_list)
+    except ValueError as e:
+        sys.exit("invalid --cnn-indexes argument " + args.cnn_indexes + str(e))
+
+    if (num_hidden_layers < args.num_lstm_layers):
+        sys.exit("--num-lstm-layers : number of lstm layers has to be greater than number of layers, decided based on splice-indexes")
+
+    # write the files used by other scripts like steps/nnet3/get_egs.sh
+    f = open(args.config_dir + "/vars", "w")
+    print('model_left_context=' + str(left_context), file=f)
+    print('model_right_context=' + str(right_context), file=f)
+    print('num_hidden_layers=' + str(num_hidden_layers), file=f)
+    # print('initial_right_context=' + str(splice_array[0][-1]), file=f)
+    f.close()
+
+    config_lines = {'components':[], 'component-nodes':[]}
+
+    config_files={}
+    prev_layer_output = nodes.AddInputLayer(config_lines, args.feat_dim, splice_indexes[0], args.ivector_dim)
+
+    # Add the init config lines for estimating the preconditioning matrices
+    init_config_lines = copy.deepcopy(config_lines)
+    init_config_lines['components'].insert(0, '# Config file for initializing neural network prior to')
+    init_config_lines['components'].insert(0, '# preconditioning matrix computation')
+    nodes.AddOutputNode(init_config_lines, prev_layer_output)
+    config_files[args.config_dir + '/init.config'] = init_config_lines
+
+    if args.ivector_dim > 0:
+      iv_layer_output = {'descriptor':  'ReplaceIndex(ivector, t, 0)',
+                         'dimension': args.ivector_dim}
+      iv_layer_output = nodes.AddAffineLayer(config_lines, "IV", iv_layer_output, args.ivector_dim, "")
+     
+    list = [('Offset(input, {0})'.format(n) if n != 0 else 'input') for n in splice_indexes[0]]
+    splice_descriptor = "Append({0})".format(", ".join(list))
+    cnn_input_dim = len(splice_indexes[0]) * args.feat_dim
+    prev_layer_output = {'descriptor':  splice_descriptor,
+                         'dimension': cnn_input_dim,
+                         '3d-dim': [len(splice_indexes[0]), args.feat_dim, 1],
+                         'vectorization': 'yzx'}
+
+    num_cnn_layers = len(cnn_array)
```

Comment:
In standard CNN recipes, the input has filter-bank coefficients, deltas and double-deltas. This gives appreciable gains. You could compute the deltas and double deltas using fixed convolution layers. You can supply the output of this fixed convolution component to the first CNN layer. The fixed convolution layer can have filters which compute static, delta and double delta features.